FEXHeaderUtils/FEXHeaderUtils/ScopedSignalMask.h

```diff
@@ -1,109 +1,61 @@
       }
-    private:
-      uint64_t OriginalMask{};
-      MutexType *Mutex;
-  };
-#else
-  // TODO: Doesn't block signals which may or may not cause issues.
-  template<typename MutexType, void (MutexType::*lock_fn)(), void (MutexType::*unlock_fn)()>
-  class ScopedSignalMaskWithMutexBase final {
-    public:
-
-      ScopedSignalMaskWithMutexBase(MutexType &_Mutex, [[maybe_unused]] uint64_t Mask = ~0ULL)
-        : Mutex {&_Mutex} {
-        // Lock the mutex
-        (Mutex->*lock_fn)();
-      }
-
-      // No copy or assignment possible
-      ScopedSignalMaskWithMutexBase(const ScopedSignalMaskWithMutexBase&) = delete;
-      ScopedSignalMaskWithMutexBase& operator=(ScopedSignalMaskWithMutexBase&) = delete;
 
-      // Only move
-      ScopedSignalMaskWithMutexBase(ScopedSignalMaskWithMutexBase &&rhs)
-       : Mutex {rhs.Mutex} {
-        rhs.Mutex = nullptr;
-      }
+      // Move-only type
+      ScopedSignalMasker(const ScopedSignalMasker&) = delete;
+      ScopedSignalMasker& operator=(ScopedSignalMasker&) = delete;
+      ScopedSignalMasker(ScopedSignalMasker &&rhs) = default;
 
-      ~ScopedSignalMaskWithMutexBase() {
-        if (Mutex != nullptr) {
-          // Unlock the mutex
-          (Mutex->*unlock_fn)();
+      ~ScopedSignalMasker() {
+        if (OriginalMask) {
+          ::syscall(SYS_rt_sigprocmask, SIG_SETMASK, &OriginalMask, nullptr, sizeof(*OriginalMask));
         }
       }
     private:
-      MutexType *Mutex;
+      std::optional<uint64_t> OriginalMask{};
   };
-
 #endif
 
-  using ScopedSignalMaskWithMutex = ScopedSignalMaskWithMutexBase<std::mutex, &std::mutex::lock, &std::mutex::unlock>;
-  using ScopedSignalMaskWithSharedLock = ScopedSignalMaskWithMutexBase<std::shared_mutex, &std::shared_mutex::lock_shared, &std::shared_mutex::unlock_shared>;
-  using ScopedSignalMaskWithUniqueLock = ScopedSignalMaskWithMutexBase<std::shared_mutex, &std::shared_mutex::lock, &std::shared_mutex::unlock>;
+  /**
+   * @brief Produces a wrapper object around a scoped lock of the given mutex
+   * while ensuring POSIX signals are masked while the mutex is locked
+   *
+   * Use this to prevent reentrancy issues of C++ mutexes with certain signal handlers.
+   * Common examples of such issues are:
+   * - C++ mutexes not unlocking due to a signal handler calling longjmp from within a scope owning the mutex
+   * - The signal handler itself using a mutex that would be re-locked if the handler gets invoked
+   *   again before unlocking
+   *
+   * Ownership of the returned object may be moved, but it is NOT SAFE to move across threads.
+   */
+  template<template<typename> class LockType = std::unique_lock, typename MutexType>
+  [[nodiscard]] static auto MaskSignalsAndLockMutex(MutexType& mutex, uint64_t Mask = ~0ULL) {
+#ifndef _WIN32
+      // Signals are masked first, and then the lock is acquired
+      return std::pair { ScopedSignalMasker { Mask }, LockType<MutexType> { mutex } };
```

Comment:
As found during development of #2739, `std::pair` has unspecified order of destruction and hence is not safe to use here.